### PR TITLE
ENH first pass at updated gfortran for osx

### PR DESCRIPTION
**Pull request by beckermr**
This recipe is a first pass at a new build of gfortran for osx. 

@mingwandroid for viz

TODO:
- [x] get the build to pass :(
- [x] Send a PR to https://github.com/conda-forge/conda-forge-repodata-patches-feedstock to fix the libgfortran dependency of existing packages to `>=3.0.0,<4.0.0.a0` (done, see https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/7)
- [x] Send a PR to https://github.com/AnacondaRecipes/aggregate/blob/master/gfortran-feedstock to fix the `run_exports` there. (done, see https://github.com/AnacondaRecipes/aggregate/pull/115)
- [x] After updated gfortran compilers are uploaded to conda-forge, rebuild conda-forge-repodata-patches (done, https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/9)
- [x] Send a PR to https://github.com/conda-forge/conda-forge-pinning-feedstock to make `fortran_compiler` use `gfortran 4.*` (done, see https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/174)
- [x] Merge gfortran-dev and libgfortran recipe to one recipe using multiple outputs
- [x] Rename gfortran-dev to gfortran_impl
- [x] Rename gfortran to gfortran_osx-64
- [x] Send a PR adjusting the run exports for the toolchain_fort package (done, https://github.com/conda-forge/toolchain-feedstock/pull/55)

### CODE REVIEW

**Comment by beckermr** [comment]
I've included the libgfortran and gfortran recipes in this PR for testing, but presumably those would be moved to other feedstocks.

**Comment by beckermr** [comment]
Currently the gfortran-dev build's locally on both compiler sets.

**Comment by beckermr** [comment]
@conda-forge/core Can we start discussing this PR and the steps needed to merge this updated tooling into conda-forge?

Here is our discussion of this previously for reference: https://github.com/conda-forge/libgfortran-feedstock/pull/16

**Comment by isuruf** [comment]
Added a TODO

**Comment by beckermr** [comment]
Thank you @isuruf!

One thing I meant to bring up is that the builds here are not full bootstrap builds. I don't have enough expertise to know exactly what this means, but the code is being built with a production version of clang. The rationale for the bootstrap builds IIUIC is to generate better compilers. Given that clang is already pretty good, I am wondering how necessary a bootstrap build really is.

**Comment by isuruf** [comment]
@beckermr, I don't think the compiler is written in Fortran, so you don't need to rebuild the compiler.

**Comment by beckermr** [comment]
I would assume the compiler would need to be rebuilt in order to compiler the new language features in FORTAN2008, but I don't know a lot here.

**Comment by isuruf** [comment]
I think you misunderstood me. Say you want to compile gcc8 which is written C. So, you get an old C compiler, say gcc4 and you compile gcc8. Programs compiled by the new gcc8 run fast, but gcc8 itself runs slow because it was copiled by gcc4 which didn't have optimizations introduced in gcc8. So you recompile gcc8 using your newly built gcc8 so that gcc8 itself runs fast. This rebuild doesn't impact the programs built, but only improves the performance of the compiler.

I'm assuming that gfortran is not written in Fortran and that only in C. Therefore if you use a good C compiler, gfortran itself will run fast and you don't need to rebuild it.

**Comment by beckermr** [comment]
Ahhhhhhh yes! So the boostrap build is not needed here like I thought. Thanks and that makes a ton of sense!

**Comment by beckermr** [comment]
OK @isuruf, I have completed the list above. I linked the PRs so they are easy to keep track of. Shall we start merging them?

**Comment by beckermr** [comment]
@isuruf Do we need to adjust this version string too?

https://github.com/conda-forge/toolchain-feedstock/blob/master/recipe/meta.yaml#L46

**Comment by beckermr** [comment]
This PR has been rebased on master.

**Comment by beckermr** [comment]
Aaaaand now I am seeing a really weird build failure. It cannot find library `-ldylib1.10.5.o` which looks like some gobbly gook.

**Comment by beckermr** [comment]
As usual, this passes locally.

**Comment by beckermr** [comment]
Somebody somewhere is setting the macosx min version without our consent. 

```
$SRC_DIR/build_conda/./gcc/xgcc -B$SRC_DIR/build_conda/./gcc/ -B$PREFIX/x86_64-apple-darwin13.4.0/bin/ -B$PREFIX/x86_64-apple-darwin13.4.0/lib/ -isystem $PREFIX/x86_64-apple-darwin13.4.0/include -isystem $PREFIX/x86_64-apple-darwin13.4.0/sys-include    -O2  -g -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -I$PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/gfortran_impl_osx-64-7.3.0 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -DIN_GCC    -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wstrict-prototypes -Wmissing-prototypes -Wold-style-definition  -isystem ./include   -mmacosx-version-min=10.5 -pipe -fPIC -g -DIN_LIBGCC2 -fbuilding-libgcc -fno-stack-protector  -dynamiclib -nodefaultlibs -install_name $PREFIX/lib/libgcc_s.1.dylib -single_module -o ./libgcc_s.dylib -Wl,-exported_symbols_list,libgcc.map -compatibility_version 1 -current_version 1.0 -g -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -I$PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/gfortran_impl_osx-64-7.3.0 -fdebug-prefix-```

File: recipes/gfortran_impl_osx-64/build.sh [hunk 1]
```diff
@@ -5,9 +5,8 @@ set -e
 mkdir -p ${SRC_DIR}/build_conda
 cd ${SRC_DIR}/build_conda
 
-export MACOSX_DEPLOYMENT_TARGET="10.9"
-
-env
+export CFLAGS="${CFLAGS} -mmacosx-version-min=${MACOSX_VERSION_MIN}"
+export LDFLAGS="${LDFLAGS} -mmacosx-version-min=${MACOSX_VERSION_MIN}"
```
Comment (beckermr) [inline recipes/gfortran_impl_osx-64/build.sh]: Thanks @isuruf! I found a hard coded value in the build scripts for libgcc at 10.5 as well. I am going to push a patch.

Comment (beckermr) [reply recipes/gfortran_impl_osx-64/build.sh]: The build fails locally with these extra lines. It is the LDFLAGS line. I removed it.

**Comment by beckermr** [comment]
I've been digging through the build scripts and headers. They seem to assume the min deployment target is 10.5. I wonder if we need to build with this instead of specifying 10.9 in some places. I am not sure if we can force the gfortran build to use 10.9

File: recipes/gfortran_impl_osx-64/build.sh (hunk 1 above)
Comment (beckermr) [reply recipes/gfortran_impl_osx-64/build.sh]: The CFLAGS change broke the build too. :(

**Comment by beckermr** [comment]
Wooooo! @isuruf let's go ahead with rest of the PRs!

File: recipes/gfortran_impl_osx-64/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,104 @@
+{% set gfortran_version = [7, 3, 0] %}
+{% set libgfortran_version = [4, 0, 0] %}
+{% set max_libgfortran_version = "5.0.0.a0" %}
+{% set libquadmath_version = [0, 0, 0] %}
+{% set libgcc_s_version = [1, 0, 0] %}
+
+package:
+  name: gfortran_impl_osx-64
```
Comment (beckermr) [inline recipes/gfortran_impl_osx-64/meta.yaml]: @isuruf I just realized I put the system postfix here at the top level package name. Did you want this or only `gfortran_impl`?

File: recipes/gfortran_osx-64/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    - lib/gcc/{{ chost }}/{{ gfortran_version|join('.') }}/finclude/**
+
+    # include{,-fixed} may not be needed unless -fopenmp is passed (not sure on that):
+    - lib/gcc/{{ chost }}/{{ gfortran_version|join('.') }}/include-fixed/**
+    - lib/gcc/{{ chost }}/{{ gfortran_version|join('.') }}/include/**
+
+    # Stub libraries:
+    - lib/libgcc_ext.10.4.dylib
+    - lib/libgcc_ext.10.5.dylib
```
Comment (isuruf) [inline recipes/gfortran_osx-64/meta.yaml]: `gfortran_osx-64` should just depend on `gfortran_impl_osx-64` at runtime. This set of files should be moved to `gfortran_osx-64` as the set of files as in https://conda.io/projects/conda-build/en/latest/source/define-metadata.html#specifying-files-to-include-in-output

Comment (beckermr) [reply recipes/gfortran_osx-64/meta.yaml]: Do you mean only the dylibs, or do we want gfortran_osx-64 to be basically a shell package?

Comment (isuruf) [reply recipes/gfortran_osx-64/meta.yaml]: I meant all files. We want `gfortran_osx-64` to be a package with just activation scripts like `gcc_osx-64`

**Comment by beckermr** [comment]
There are a few issues here:

1. To build gcc/gfortran right, you need to install it with the build system, e.g. `make install-strip`. I was using the `gfortran_imp_osx-64` output prefix to hold this, but this includes a bunch of gcc stuff we don't want. I don't want to copy these files out of the work dir directly because IDK what exactly the gcc build is doing when running `make install-strip`.

2. I don't really understand what the docs mean for what `outputs/files:` does. They say

> Explicit file lists are relative paths from the root of the build prefix.

Some outputs may not have a build prefix. Sometimes this seems to be the work dir for the original build maybe or the host prefix? Is this the top level build prefix? Is the top level package the only one that is supposed to have a build prefix?

3. I even had to hack around listing build and host sections in outputs. At one point it was putting a package that I listed in the build section of an output into a host prefix for that output. 

Given all of this, I am trying the following:

1. I am making a directory in the work dir which I feed to configure as the installation prefix.
2. I am then writing scripts for each output that copy the files to the right spot from this work dir.

I am more than happy to do something else. This simply seemed more robust.

File: .ci_support/compute_build_graph.py [hunk 4]
```diff
@@ -353,7 +353,9 @@ def collapse_subpackage_nodes(graph):
             if subpackages:
                 remap_edges = [edge for edge in graph.edges() if edge[1] in subpackages]
                 for edge in remap_edges:
-                    graph.add_edge(edge[0], master_key)
+                    # make sure not to add references to yourself
+                    if edge[0] != master_key:
+                        graph.add_edge(edge[0], master_key)
```
Comment (beckermr) [inline .ci_support/compute_build_graph.py]: @isuruf @conda-forge/core This is a bug fix for build recipes with multiple outputs that have a specific dependency order.

**Comment by beckermr** [comment]
Friendly bump on this @conda-forge/core @mingwandroid!

We need to merge this PR on the upstream recipe: https://github.com/AnacondaRecipes/aggregate/pull/115

File: recipes/gfortran_osx-64/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  license: GPL 3 (with GCC Runtime Library Exception 3.1)
+  license_family: GPL
+  summary: Fortran libraries from the GNU Compiler Collection
+
+extra:
+  recipe-maintainers:
+    - jakirkham
+    - msarahan
+    - pelson
+    - beckermr
```
Comment (scopatz) [inline recipes/gfortran_osx-64/meta.yaml]: @msarahan @pelson @jakirkham - Are you all OK with being maintainers here?

File: .travis.yml [hunk 6]
```diff
@@ -43,6 +43,5 @@ script:
         source ./.travis_scripts/create_feedstocks;
       else
         echo "Building all recipes.";
-        source ./.travis_scripts/build_all;
+        travis_wait 60 source ./.travis_scripts/build_all;
```
Comment (beckermr) [inline .travis.yml]: This needs to be removed before merging. On the upstream feedstock, we will have to add this fix back I think. Or maybe switch to a circleci osx build and let the build be more noisy.

File: recipes/gfortran_osx-64/meta.yaml (hunk 5 above)
Comment (beckermr) [reply recipes/gfortran_osx-64/meta.yaml]: Bump on this friends!

**Comment by beckermr** [comment]
Thanks for the fix @isuruf!

File: recipes/gfortran_osx-64/meta.yaml (hunk 5 above)
Comment (beckermr) [reply recipes/gfortran_osx-64/meta.yaml]: Shall we remove the other maintainers and merge @isuruf @scopatz?

Core will have access no matter what and other ppl can be added as needed.

File: recipes/gfortran_impl_osx-64/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - make
+  host:
+    - libiconv {{ libiconv }}
+    - zlib {{ zlib }}
+    - gmp {{ gmp }}
+    - mpfr {{ mpfr }}
+    - isl {{ isl }}
+    - mpc ==1.1.0
```
Comment (isuruf) [inline recipes/gfortran_impl_osx-64/meta.yaml]: Can you remove the pinnings on the 6 packages above?

File: recipes/gfortran_impl_osx-64/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+      host:
+        - libiconv {{ libiconv }}
+        - zlib {{ zlib }}
+        - gmp {{ gmp }}
+        - mpfr {{ mpfr }}
+        - isl {{ isl }}
+        - mpc ==1.1.0
```
Comment (isuruf) [inline recipes/gfortran_impl_osx-64/meta.yaml]: Same here

File: recipes/gfortran_impl_osx-64/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+      run:
+        - libgfortran >={{ libgfortran_version|join('.') }},<{{ max_libgfortran_version }}
+        - {{ pin_compatible('isl', max_pin='x.x') }}
+        - {{ pin_compatible('mpfr', max_pin='x.x') }}
+        - {{ pin_compatible('mpc', max_pin='x.x') }}
+        - {{ pin_compatible('gmp', max_pin='x.x') }}
+        - {{ pin_compatible('zlib', max_pin='x.x') }}
+        - {{ pin_compatible('libiconv', max_pin='x.x') }}
```
Comment (isuruf) [inline recipes/gfortran_impl_osx-64/meta.yaml]: Same here

File: recipes/gfortran_osx-64/run_test.sh [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+# we need to tell the linker where to look in our tests
+# this step is not needed when using conda build
+export DYLD_FALLBACK_LIBRARY_PATH=${PREFIX}/lib
```
Comment (isuruf) [inline recipes/gfortran_osx-64/run_test.sh]: I don't understand why this is needed? Isn't `rpath` set in the activation scripts?

Comment (beckermr) [reply recipes/gfortran_osx-64/run_test.sh]: This was in the old recipe. In my experience it has failed without. I think the reason is that the executable produced by the compiler itself is not run through conda build. This could not be true anymore.

File: recipes/gfortran_impl_osx-64/meta.yaml (hunk 7 above)
Comment (beckermr) [reply recipes/gfortran_impl_osx-64/meta.yaml]: These pick up the central conda forge pinnings. Is that not what we want?

Happy to remove them! Not at my machine right now though.

Comment (isuruf) [reply recipes/gfortran_impl_osx-64/meta.yaml]: You can replace `- isl {{ isl }}` with just `- isl`. It does the same thing

File: recipes/gfortran_osx-64/run_test.sh (hunk 10 above)
Comment (isuruf) [reply recipes/gfortran_osx-64/run_test.sh]: @mingwandroid, which package sets LDFLAGS in osx? Is it the C compiler?

File: recipes/gfortran_impl_osx-64/meta.yaml (hunk 7 above)
Comment (beckermr) [reply recipes/gfortran_impl_osx-64/meta.yaml]: Ahhh got it. Thanks!

File: recipes/gfortran_osx-64/run_test.sh (hunk 10 above)
Comment (mingwandroid) [reply recipes/gfortran_osx-64/run_test.sh]: clang_osx-64 AFAIR. 

> `export DYLD_FALLBACK_LIBRARY_PATH`

.. should probably be removed. It does nothing on machines with SIP enabled. If it cannot be built with SIP enabled it's unlikely that is will build on ci systems, which I assume enabled SIP. What do the MacOS ci run, MacOS version wise?

Comment (mingwandroid) [reply recipes/gfortran_osx-64/run_test.sh]: clang_osx-64 AFAIR. 

> `export DYLD_FALLBACK_LIBRARY_PATH`

.. should probably be removed. It does nothing on machines with SIP enabled. If it cannot be built with SIP enabled it's unlikely that is will build on ci systems, which I assume enabled SIP. What do the MacOS ci run, MacOS version wise?

Comment (beckermr) [reply recipes/gfortran_osx-64/run_test.sh]: This has been building fine on Travis. Happy to remove.

Comment (beckermr) [reply recipes/gfortran_osx-64/run_test.sh]: We use Xcode 9.4 and OS X 10.13

**Comment by beckermr** [comment]
Thanks @isuruf !

File: recipes/gfortran_osx-64/run_test.sh [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+gfortran -o hello hello.f90 -Wl,-rpath,${PREFIX}/lib
```
Comment (isuruf) [inline recipes/gfortran_osx-64/run_test.sh]: Can you change this to `$FC`?

Comment (isuruf) [reply recipes/gfortran_osx-64/run_test.sh]: Can you also add a test to see if `FFLAGS` is non-empty?

File: recipes/gfortran_osx-64/run_test.sh (hunk 10 above)
Comment (mingwandroid) [reply recipes/gfortran_osx-64/run_test.sh]: What does `csrutil status` print? Would also be keen to find out on azurepipelines.

**Comment by beckermr** [comment]
Ok.  I think I got everything. Let’s see if this passes.

**Comment by beckermr** [comment]
Alright. Tests pass! Back to y'all @isuruf @scopatz!

**Comment by beckermr** [comment]
OK. I have removed the hack on travis and the other maintainers. Shall we merge this @isuruf @scopatz ?

**Comment by scopatz** [comment]
@isuruf is studying for quals, so I am happy with this once CI passes

**Comment by beckermr** [comment]
Ohhh boy @isuruf! Good luck! I did that ages ago at UChicago. That was not fun. I skipped the CI since the changes I made were not to the build. I forgot to skip azure. :(

File: recipes/gfortran_impl_osx-64/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+      - lib/libgcc_ext.10.4.dylib
+      - lib/libgcc_ext.10.5.dylib
+
+    test:
+      script: run_test.sh
+
+about:
+  home: http://gcc.gnu.org/
+  license: GPL 3 (with GCC Runtime Library Exception 3.1)
+  license_family: GPL
```
Comment (scopatz) [inline recipes/gfortran_impl_osx-64/meta.yaml]: You need to have the license file here for GPL licenses

Comment (beckermr) [reply recipes/gfortran_impl_osx-64/meta.yaml]: Frack. Good catch!

Comment (beckermr) [reply recipes/gfortran_impl_osx-64/meta.yaml]: I am surprised the build passed...

**Comment by beckermr** [comment]
I added a copy of the license file to the recipe that did not have the source tarball.

**Comment by beckermr** [comment]
Mmmk. Back to you @scopatz !

**Comment by beckermr** [comment]
huzzzah!

**Comment by scopatz** [comment]
:tada:

**Comment by beckermr** [comment]
I will probably have to make a PR on the feedstock once it is built in order for the first build to work. Travis will likely time out.
